appserver/microprofile/health-glassfish/src/main/java/org/glassfish/microprofile/impl/CollectHealthChecksExtension.java

```diff
@@ -0,0 +1,75 @@
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+package org.glassfish.microprofile.impl;
+
+import jakarta.enterprise.context.spi.CreationalContext;
+import jakarta.enterprise.event.Observes;
+import jakarta.enterprise.inject.spi.AfterDeploymentValidation;
+import jakarta.enterprise.inject.spi.Annotated;
+import jakarta.enterprise.inject.spi.Bean;
+import jakarta.enterprise.inject.spi.BeanManager;
+import jakarta.enterprise.inject.spi.Extension;
+import jakarta.enterprise.inject.spi.ProcessBean;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.microprofile.health.HealthCheck;
+import org.eclipse.microprofile.health.Liveness;
+import org.eclipse.microprofile.health.Readiness;
+import org.eclipse.microprofile.health.Startup;
+import org.glassfish.hk2.api.ServiceLocator;
+import org.glassfish.hk2.utilities.ServiceLocatorUtilities;
+import org.glassfish.internal.api.Globals;
+import org.glassfish.microprofile.health.HealthReporter;
+
+
+public class CollectHealthChecksExtension implements Extension {
+
+    private final HealthReporter service;
+    private final Set<Bean<HealthCheck>> healthChecks = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+    public CollectHealthChecksExtension() {
+        ServiceLocator defaultBaseServiceLocator = Globals.getDefaultBaseServiceLocator();
+        HealthReporter healthReporterService = defaultBaseServiceLocator.getService(HealthReporter.class);
+        if (healthReporterService == null) {
+            ServiceLocatorUtilities.addClasses(defaultBaseServiceLocator, true, HealthReporter.class);
+            healthReporterService = defaultBaseServiceLocator.getService(HealthReporter.class);
+        }
+        service = healthReporterService;
+    }
+    public <T> void processBeans(@Observes ProcessBean<T> beans) {
+        Annotated annotated = beans.getAnnotated();
+        if (annotated.isAnnotationPresent(Liveness.class) ||
+                annotated.isAnnotationPresent(Readiness.class) ||
+                annotated.isAnnotationPresent(Startup.class)) {
+            if (beans.getBean().getTypes().contains(HealthCheck.class)) {
+                healthChecks.add((Bean<HealthCheck>) beans.getBean());
+            }
+        }
+    }
+
+    public void afterDeploymentValidation(@Observes AfterDeploymentValidation adv, BeanManager beanManager) {
+        healthChecks.forEach(bean -> {
+            CreationalContext<HealthCheck> creationalContext = beanManager.createCreationalContext(bean);
+            service.addHealthCheck("", bean.create(creationalContext));
```

Comment:
This should bind the health checks to application name, which has to be unique. When undeployed, `HealthService.event` should unbind health checks for the same application name.

The problem with the current approach is that when 2 applications are deployed, they deploy all health checks under the same context name. When one of the applications is undeployed, it would remove health checks also from other applications. We need to add and remove health checks for a single application under a unique context.

Reply:
From within an application. e.g. in the CDI extension, you can get the appname from InvocationManager.

There are multiple ways to retrieve it. The most straightforward is to get it from the global HK2 service locator via Globals.get(InvocationManager.class).

On InvocationManager, you get the app name:

```
InvocatoinManager im;
im.getCurrentInvocation().getAppName();
```